chainer/function_node.py

```diff
@@ -0,0 +1,590 @@
+
+           x.node <--- f <--- y.node
+
+       If an application of another function ``g`` occurs as
+
+       >>> g = F.Identity()
+       >>> z = g.apply((x,))[0]
+
+       then the graph grows with a branch::
+
+                    |--- f <--- y.node
+           x.node <-+
+                    |--- g <--- z.node
+
+       Note that the branching is correctly managed on backward computation,
+       i.e. the gradients from ``f`` and ``g`` are accumulated to the gradient
+       of ``x``.
+
+    Every function-node implementation should provide :meth:`forward` and
+    :meth:`backward`. Instead of overriding :meth:`forward`, one can also
+    implement :meth:`forward_cpu` and :meth:`forward_gpu` when the
+    implementations for CPU and GPU arrays are totally different.
+
+    Note that the input and output variables are inaccessible from
+    :meth:`backward` by default. If it needs accesses to these variables, the
+    :meth:`forward` method (or its CPU/GPU variants) has to call
+    :meth:`retain_inputs` and :meth:`retain_outputs` appropriately. The
+    retained input/output variables can be accessed from :meth:`backward` by
+    calling :meth:`get_retained_inputs` and :meth:`get_retained_outputs`.
+
+    .. note::
+
+       There are two types of differentiable functions in Chainer (since v3).
+       The first type is of a function using a subclass of :class:`Function`,
+       which is called *old-style differentiable function*. The second type is
+       of a function using a subclass of :class:`FunctionNode`, which is called
+       **new-style differentiable function**. There are several advantages on
+       using the new-style differentiable function.
+
+       - The new-style differentiable function supports *differentiable
+         backpropagation*. The backpropagated gradients computed through the
+         new-style differentiable functions themselves support further
+         backpropagations so that the automatic higher-order differentiation is
+         available.
+       - The backpropagation of the new-style differentiable function can be
+         more computationally saving because the interface allows an
+         implementation to omit the computation of unneeded input gradients.
+
+       Note that the new-style differentiable function is the standard way of
+       defining a function node of the computational graph in Chainer; old-
+       style differentiable functions are implemented as wrappers of the new-
+       style differentiable functions.
+
+    Attributes:
+        inputs: A tuple of the input :class:`VariableNode` objects.
+        outputs: A tuple of weak references to the output
+            :class:`VariableNode` objects.
+        rank (int): An ordinal following the topological order of the
+            computational graph.
+        stack: Stack trace retrieved at the forward computation. The stack
+            trace is available only in the debug mode.
```

Comment:
I think there is no good pointer to debug mode in the reference now. Is it enough to just add a link to configuration.rst?